urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/MuxClient.java

```diff
@@ -0,0 +1,490 @@
+						if (seal != null) {
+							// AES encryption
+							final byte[] encoded = IOHelper.fromWireWithHeader(is);
+							final byte[] decoded = seal.decode(encoded, 0, encoded.length);
+							final ByteArrayInputStream bais = new ByteArrayInputStream(decoded);
+							msg.fromWire(bais);
+						} else {
+							msg.fromWire(is);
+						}
+						context.getStatistics().incInMsgs().incInBytes(msg.getBufferLen());
+						router.onReceiveFromRemote(this, msg);
+						context.releaseMuxPacket(msg);
+					} catch (SocketTimeoutException e) {
+						final long now = System.currentTimeMillis();
+						if ((muxNopKeepAlive + Constants.MUX_KEEP_ALIVE) < now) {
+							Log.debug(this.getClass().getSimpleName() + " " + e.toString());
+							sendNOP();
+							muxNopKeepAlive = now;
+						}
+						continue;
+					} catch (EOFException e) {
+						Log.error(this.getClass().getSimpleName() + " " + e.toString());
+						doSleep(1000);
+						break;
+					} catch (IOException e) {
+						if (!sock.isClosed() && !shutdown) {
+							Log.error(this.getClass().getSimpleName() + " " + e.toString());
+						}
+						break;
+					} catch (GeneralSecurityException e) {
+						Log.error(this.getClass().getSimpleName() + " " + e.toString());
+						doSleep(1000);
+						break;
+					} catch (Exception e) {
+						Log.error(this.getClass().getSimpleName() + " Generic exception", e);
+						break;
+					}
+				}
+				// Close all
+				Log.info(this.getClass().getSimpleName() + " close all: " + sock);
+				close();
+				synchronized (mapLocals) { // Locals are RAW
+					for (MuxClientLocal l : mapLocals.values()) {
+						l.setShutdown();
+					}
+					mapLocals.clear();
+				}
+			}
+			Log.info(this.getClass().getSimpleName() + " await end");
+			context.awaitShutdown(this);
+			Log.info(this.getClass().getSimpleName() + " end");
+		}
+	}
+
+	class MuxClientLocal extends MuxClientConnection { // Local is RAW
+		final Semaphore isLocked = new Semaphore(Constants.BUFFER_LEN * Constants.IO_BUFFERS);
+		final ArrayBlockingQueue<RawPacket> queue = new ArrayBlockingQueue<RawPacket>(
+				Constants.IO_BUFFERS << 1);
+		int id;
+		InetAddress stickyAddress = null;
+		long keepalive = System.currentTimeMillis();
+
+		MuxClientLocal(final OutboundAddress outboundAddress) {
+			super(outboundAddress);
+		}
+
+		public void setSticky(final InetAddress stickyAddress) {
+			this.stickyAddress = stickyAddress;
+		}
+
+		void unlock(final int size) {
+			isLocked.release(size);
+		}
+
+		boolean lock(final int size) throws InterruptedException {
+			return isLocked.tryAcquire(size, 3000, TimeUnit.MILLISECONDS);
+		}
+
+		void setId(final int id) {
+			this.id = id;
+		}
+
+		void sendQueueLocal(final RawPacket msg) throws IOException {
+			try {
+				while (!queue.offer(msg, 1000, TimeUnit.MILLISECONDS)) {
+					if (shutdown)
+						break;
+				}
+				keepalive = System.currentTimeMillis();
+			} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)